llama_index/query_engine/retry_query_engine.py

```diff
@@ -0,0 +1,292 @@
+        )
+
+        service_context = service_context or ServiceContext.from_defaults()
+
+        return cls(
+            retriever=retriever,
+            service_context=service_context,
+            response_synthesizer=response_synthesizer,
+            callback_manager=callback_manager,
+            check_source=check_source,
+            percentile_cutoff=percentile_cutoff,
+            use_shrinking_percentile_cutoff=use_shrinking_percentile_cutoff,
+            max_retries=max_retries,
+        )
+
+    def retrieve(self, query_bundle: QueryBundle) -> List[NodeWithScore]:
+        return self._retriever.retrieve(query_bundle)
+
+    def synthesize(
+        self,
+        query_bundle: QueryBundle,
+        nodes: List[NodeWithScore],
+        additional_source_nodes: Optional[Sequence[NodeWithScore]] = None,
+    ) -> RESPONSE_TYPE:
+        return self._response_synthesizer.synthesize(
+            query_bundle=query_bundle,
+            nodes=nodes,
+            additional_source_nodes=additional_source_nodes,
+        )
+
+    async def asynthesize(
+        self,
+        query_bundle: QueryBundle,
+        nodes: List[NodeWithScore],
+        additional_source_nodes: Optional[Sequence[NodeWithScore]] = None,
+    ) -> RESPONSE_TYPE:
+        return await self._response_synthesizer.asynthesize(
+            query_bundle=query_bundle,
+            nodes=nodes,
+            additional_source_nodes=additional_source_nodes,
+        )
+
+    def _query(self, query_bundle: QueryBundle) -> RESPONSE_TYPE:
+        """Answer a query."""
+        query_id = self.callback_manager.on_event_start(CBEventType.QUERY)
+
+        retrieve_id = self.callback_manager.on_event_start(CBEventType.RETRIEVE)
+        nodes = self._retriever.retrieve(query_bundle)
+        self.callback_manager.on_event_end(
+            CBEventType.RETRIEVE, payload={"nodes": nodes}, event_id=retrieve_id
+        )
+
+        synth_id = self.callback_manager.on_event_start(CBEventType.SYNTHESIZE)
+        response = self._response_synthesizer.synthesize(
+            query_bundle=query_bundle,
+            nodes=nodes,
+        )
+        self.callback_manager.on_event_end(
+            CBEventType.SYNTHESIZE, payload={"response": response}, event_id=synth_id
+        )
+
+        self.callback_manager.on_event_end(CBEventType.QUERY, event_id=query_id)
+
+        # Insert retry here
+        if self.max_retries > 0:
+            evaluator = QueryResponseEvaluator(service_context=self.service_context)
+
+            if type(response) is StreamingResponse:
+                typed_response = response.get_response()
+            elif type(response) is Response:
+                typed_response = response
+            else:
+                raise ValueError(f"Unexpected response type {type(response)}")
+
+            response_eval = evaluator.evaluate(query_bundle.query_str, typed_response)
+            if response_eval == "YES":
```

Comment:
i think this could be generalized a bit into an overall "evaluator" function (we don't have the proper abstractions for this yet, to start with this could just be a Callable function). The current evaluator outputs aren't the most abstract - they return a "YES" or "NO" string. Could we have a configurable evaluator Calalbe that can be passed in the `__init__`? It can use `QueryResponseEvaluator` as a default